zio-logback-http4s/src/main/scala/com/softwaremill/correlator/CorrelationIdMiddleware.scala

```diff
@@ -0,0 +1,104 @@
+
+  final val defaultHeaderName: String = "X-Correlation-ID"
+
+  final val defaultLogStartRequest: (String, Request[Task]) => Task[Unit] = (cid, req) =>
+    Task(logger.debug(s"Starting request with id: $cid, to: ${req.uri.path}"))
+
+  final val defaultIdGenerator: URIO[Random, String] =
+    ZIO.accessM[Random] { random =>
+      val randomUpperCaseChar: UIO[Char] = random.random.nextInt(91 - 65).map(r => (r + 65).toChar)
+      val segment: UIO[String]           = (1 to 3).toList.traverse(_ => randomUpperCaseChar).map(_.mkString)
+      (segment, segment, segment).mapN((a, b, c) => s"$a-$b-$c")
+    }
+
+  final val getCorrelationId: Task[Option[String]] = Task(Option(MDC.get(MdcKey)))
+
+  @silent("parameter value zioMDCAdapter in method addTo is never used")
+  def addTo[R <: Random](
+    headerName: String = defaultHeaderName,
+    logStartRequest: (String, Request[Task]) => Task[Unit] = defaultLogStartRequest,
+    idGenerator: RIO[R, String] = defaultIdGenerator
+  )(service: HttpRoutes[Task])(implicit runtime: Runtime[R], zioMDCAdapter: ZioMDCAdapter): HttpRoutes[Task] =
+    Kleisli { req: Request[Task] =>
+      val cidM: Task[String] =
+        req.headers.get(CaseInsensitiveString(headerName)).fold(idGenerator.asInstanceOf[Task[String]])(_.value.pure[Task])
+
+      val setupAndService: Task[Option[Response[Task]]] =
+        for {
+          cid <- cidM
+          _   <- ZIO(MDC.put(MdcKey, cid))
+          _   <- logStartRequest(cid, req)
+          r   <- service(req).value
+        } yield r
+
+      OptionT(setupAndService.ensuring(Task(MDC.remove(MdcKey)).orElse(URIO.unit)))
+    }
+}
+
+/**
+ * Based on [[https://olegpy.com/better-logging-monix-1/]]. Makes the current correlation id available for logback
+ * loggers.
+ */
+final class ZioMDCAdapter private[correlator] (fiber: FiberRef[ju.Map[String, String]])(implicit runtime: Runtime[_])
+    extends LogbackMDCAdapter {
+  override def get(key: String): String = runtime.unsafeRun { fiber.get.map(_.get(key)) }
+
+  override def put(key: String, `val`: String): Unit =
+    runtime.unsafeRun {
+      fiber.modify { map =>
+        map.put(key, `val`)
+        ((), map)
+      }
+    }
+
+  override def remove(key: String): Unit =
+    runtime.unsafeRun {
+      fiber.modify { map =>
+        map.remove(key)
+        ((), map)
+      }
+    }
+
+  override def clear(): Unit = runtime.unsafeRun { fiber.set(new ju.HashMap()) }
+
+  override def getCopyOfContextMap: ju.HashMap[String, String] = runtime.unsafeRun { fiber.get.map(new ju.HashMap(_)) }
+
+  override def setContextMap(contextMap: ju.Map[String, String]): Unit =
+    runtime.unsafeRun { fiber.set(new ju.HashMap(contextMap)) }
+
+  override def getPropertyMap: ju.Map[String, String] = runtime.unsafeRun { fiber.get }
+
+  override def getKeys: ju.Set[String] = runtime.unsafeRun { fiber.get.map(_.keySet()) }
+}
+
+object ZioMDCAdapter {
+  def init(implicit runtime: Runtime[_]): Task[ZioMDCAdapter] =
```

Comment:
No need to require the runtime here, just do `ZIO.runtime` inside the body to grab the runtime.

Reply:
Good idea! Thanks :)